src/app_add_cycle.rs

```diff
@@ -0,0 +1,308 @@
+        );
+        let mut cumulative_tracked_uniques = TypeIdBuckets::<CyclePluginAssociations>::new(
+            "tracked uniques in workloads",
+            &self.type_names,
+        );
+
+        'each_workload: for (
+            _workloads,
+            AppWorkloadInfo {
+                name,
+                plugin_id,
+                signature,
+                batch_info: _,
+                type_names: _,
+            },
+        ) in cycle
+        {
+            names_checked.push(name.clone());
+
+            // can happen if a cycle has the same workload multiple times
+            if let Some(ref p) = plugin_id {
+                if plugins_added.contains(p) {
+                    // so, we don't want to add duplicate associations for them
+                    continue 'each_workload;
+                } else {
+                    plugins_added.insert(*p);
+                }
+            }
+
+            // account for update packs
+            for ((up_type, _), assoc) in signature.track_update_packed.entries() {
+                if !assoc.is_empty() {
+                    cumulative_update_packed.associate(
+                        up_type,
+                        CyclePluginAssociations {
+                            plugins: assoc,
+                            plugin_id,
+                            workload: name.clone(),
+                        },
+                    );
+                }
+            }
+            // account for tracked uniques
+            for ((tracked_type, _), assoc) in signature.track_tracked_uniques.entries() {
+                if !assoc.is_empty() {
+                    cumulative_tracked_uniques.associate(
+                        tracked_type,
+                        CyclePluginAssociations {
+                            plugins: assoc,
+                            workload: name.clone(),
+                            plugin_id: None,
+                        },
+                    );
+                }
+            }
+        }
+
+        let mut errs = Vec::<CycleCheckError>::new();
+
+        // update pack
+        for ((_, update_pack_storage_name), workloads_dependent) in
+            cumulative_update_packed.entries()
+        {
+            if workloads_dependent.len() > 1 {
+                errs.push(CycleCheckError::UpdatePackResetInMultipleWorkloads {
+                    update_pack: update_pack_storage_name,
+                    conflicts: workloads_dependent,
+                })
+            }
+        }
+        // tracked unique
+        for ((_, tracked_unique_storage_name), workloads_dependent) in
+            cumulative_tracked_uniques.entries()
+        {
+            if workloads_dependent.len() > 1 {
+                errs.push(CycleCheckError::TrackedUniqueResetInMultipleWorkloads {
+                    tracked_unique: tracked_unique_storage_name,
+                    conflicts: workloads_dependent,
+                })
+            }
+        }
```

Comment:
Perhaps some functional idioms here? No hard preference though.

```suggestion
        // update pack
        let mut errs = cumulative_update_packed
            .entries()
            .filter(|((_, _), workloads_dependent)| workloads_dependent.len() > 1)
            .map(|((_, update_pack_storage_name), workloads_dependent)|
                CycleCheckError::UpdatePackResetInMultipleWorkloads {
                    update_pack: update_pack_storage_name,
                    conflicts: workloads_dependent,
                })
            .collect::<Vec<_>>();
        // tracked unique
        errs.extend(cumulative_tracked_uniques
            .entries()
            .filter(|((_, _), workloads_dependent)| workloads_dependent.len() > 1)
            .map(|((_, tracked_unique_storage_name), workloads_dependent)|
                CycleCheckError::TrackedUniqueResetInMultipleWorkloads {
                    tracked_unique: tracked_unique_storage_name,
                    conflicts: workloads_dependent,
                })
            );
```

Reply:
Very nice!